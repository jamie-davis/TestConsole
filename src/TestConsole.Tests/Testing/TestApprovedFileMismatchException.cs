using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using NUnit.Framework;
using TestConsoleLib;
using TestConsoleLib.Exceptions;
using TestConsoleLib.Testing;

namespace TestConsole.Tests.Testing
{
    [TestFixture]
    public class TestApprovedFileMismatchException
    {
        [Test]
        public void ExceptionMessageIsFormatted()
        {
            //Arrange
            Exception exception = null;
            var approvedText = @"Line 1
Line 2";
            var receivedText = @"Line 1
Line 2 Diff";

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesEmptyReceivedFile()
        {
            //Arrange
            Exception exception = null;
            var approvedText = @"Line 1
Line 2";
            var receivedText = string.Empty;

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesEmptyApprovedFile()
        {
            //Arrange
            Exception exception = null;
            var approvedText  = string.Empty;
            var receivedText= @"Line 1
Line 2";

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void ExceptionMessageHandlesIdenticalData()
        {
            //Arrange
            Exception exception = null;
            var approvedText  = string.Empty;

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(approvedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }

        [Test]
        public void MaximumDifferencesAreReported()
        {
            //Arrange
            Exception exception = null;
            var approvedText = string.Join(Environment.NewLine, Enumerable.Range(1, 100).Select(n => $"Line {n}"));
            var receivedText = string.Join(Environment.NewLine, Enumerable.Range(1, 100).Select(n => $"Line {((n & 1) != 0 ? "X" : n.ToString())}"));

            using (var approved = WriteTempFile(approvedText))
            using (var received = WriteTempFile(receivedText))
            {
                //Act
                exception = new ApprovedFileMismatchException(received.TempPath, approved.TempPath);
            }

            //Assert
            exception.Message.Verify();
        }


        #region Temporary file handling

        private TempFile WriteTempFile(string text)
        {
            return new TempFile(text);
        }

        private class TempFile : IDisposable
        {
            public string TempPath { get; }

            public TempFile(string text)
            {
                TempPath = Path.GetTempFileName();
                try
                {
                    File.WriteAllText(TempPath, text);
                }
                catch
                {
                    Dispose(); //We will not complete construction, don't leave the temp file in place
                    throw;
                }
            }

            #region IDisposable

            public void Dispose()
            {
                File.Delete(TempPath);
            }

            #endregion
        }

        #endregion
    }
}
