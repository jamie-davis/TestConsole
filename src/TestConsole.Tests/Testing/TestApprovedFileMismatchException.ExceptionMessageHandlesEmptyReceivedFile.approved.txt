2 Differences found between approved and received:                                             
                                                                                               
File                   Length       Lines                                                      
-------- -------------------- -----------                                                      
Approved                   13           2                                                      
Received                    0           0                                                      
                                                                                               
File             Line Text                                                                     
-------- ------------ ------                                                                   
Approved            1 Line 1                                                                   
Received            1                                                                          
Approved            2 Line 2                                                                   
Received            2                                                                          
                                                                                               